nemo/collections/asr/parts/numba/rnnt_loss/utils/cpu_utils/cpu_rnnt.py

```diff
@@ -0,0 +1,378 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Copyright 2018-2019, Mingkun Huang
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import math
+import multiprocessing
+from typing import Optional
+
+import numba
+import torch
+
+from nemo.collections.asr.parts.numba.rnnt_loss.utils import global_constants
+
+
+def log_sum_exp(a: torch.Tensor, b: torch.Tensor):
+    """
+    Logsumexp with safety checks for infs.
+    """
+    if torch.isinf(a):
+        return b
+
+    if torch.isinf(b):
+        return a
+
+    if a > b:
+        return math.log1p(math.exp(b - a)) + a
+    else:
+        return math.log1p(math.exp(a - b)) + b
+
+
+class CpuRNNT_index:
+    def __init__(self, U: int, maxU: int, minibatch: int, alphabet_size: int, batch_first: bool):
+        """
+        A placeholder Index computation class that emits the resolved index in a flattened tensor,
+        mimicing pointer indexing in CUDA kernels on the CPU.
+
+        Args:
+            U: Length of the current target sample (without padding).
+            maxU: Max Length of the padded target samples.
+            minibatch: Minibatch index
+            alphabet_size: Size of the vocabulary including RNNT blank - V+1.
+            batch_first: Bool flag determining if batch index is first or third.
+        """
+        super(CpuRNNT_index, self).__init__()
+        self.U = U
+        self.maxU = maxU
+        self.minibatch = minibatch
+        self.alphabet_size = alphabet_size
+        self.batch_first = batch_first
+
+    def __call__(self, t: int, u: int, v: Optional[int] = None):
+        # if indexing all the values of the vocabulary, then only t, u are provided
+        if v is None:
+            return t * self.U + u
+        else:
+            # otherwise, t, u, v are provided to index particular value in the vocabulary.
+            if self.batch_first:
+                return (t * self.maxU + u) * self.alphabet_size + v
+            else:
+                return (t * self.maxU + u) * self.minibatch * self.alphabet_size + v
+
+
+class CpuRNNT_metadata:
```

Comment:
CPURNNT?

Reply:
Ah this is the metadata used to compute a fast index for the blank and label token tensors. The CPURNNT below uses this.

Reply:
I just realized you meant the name. I kept it 1:1 with the C++ codebase, so lets keep it as is here.